### PR TITLE
HOSTEDCP-1256: control-plane-pki-operator: add an operator for managing PKI

### PR DESCRIPTION
**Pull request by stevekuznetsov**
go.mod: update vendors

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

*: add a control-plane-pki-operator

In the future, we'd like to lean on the library-go approach for PKI
management to gain access to some of the features it provides as well as
increasing the number of engineers who are able to maintain PKI in
HyperShift as we'd no longer be using an in-house solution.

The first step in this process is adding a new controller to manage PKI
for the hosted control planes, and having it manage:

 - a new signer for customers' break-glass credentials
 - a trust bundle of active signers
 - a client certificate and key signed by the currently-valid signer
 - a total client CA bundle (which is, for now, the smae as 2)

The control-plane-operator consumes the total client CA and aggregates
it into the current KAS client CA flow.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

/assign @sjenning @deads2k

### CODE REVIEW

File: control-plane-pki-operator/certrotationcontroller/certrotationcontroller.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+) (*CertRotationController, error) {
+	ret := &CertRotationController{
+		recorder: eventRecorder,
+		cachesToSync: []cache.InformerSynced{
+			configInformer.Config().V1().Networks().Informer().HasSynced,
+			configInformer.Config().V1().Infrastructures().Informer().HasSynced,
+		},
+	}
+
+	rotationDay := defaultRotationDay // TODO(deads2k): is this logic we want to persist? looks temporary ...
```
Comment (stevekuznetsov) [inline control-plane-pki-operator/certrotationcontroller/certrotationcontroller.go:56]: @deads2k

File: control-plane-pki-operator/certrotationcontroller/certrotationcontroller.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+			EventRecorder: eventRecorder,
+			Owner:         ownerRef,
+		},
+		certrotation.RotatedSelfSignedCertKeySecret{
+			Namespace: hostedControlPlane.Namespace,
+			Name:      manifests.CustomerSystemAdminClientCertSecret(hostedControlPlane.Namespace).Name,
+			Validity:  36 * rotationDay / 24,
+			Refresh:   6 * rotationDay / 24,
+			CertCreator: &certrotation.ClientRotation{
+				UserInfo: &user.DefaultInfo{Groups: []string{"system:masters"}}, // TODO: name, uid - what to set?
```
Comment (stevekuznetsov) [inline control-plane-pki-operator/certrotationcontroller/certrotationcontroller.go:103]: @deads2k

File: control-plane-pki-operator/operator.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	"github.com/openshift/library-go/pkg/operator/configobserver/featuregates"
+	"github.com/openshift/library-go/pkg/operator/status"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+)
+
+func RunOperator(ctx context.Context, controllerContext *controllercmd.ControllerContext) error {
+	var namespace, name string
```
Comment (stevekuznetsov) [inline control-plane-pki-operator/operator.go:24]: @deads2k would like your guidance on parameterizing this - the controllercmd libarry does not allow for flags and will require hackery one way or ther other ...

File: control-plane-pki-operator/targetconfigcontroller/targetconfigcontroller.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+// createTargetConfig takes care of creation of valid resources in a fixed name.  These are inputs to other control loops.
+// returns whether or not requeue and if an error happened when updating status.  Normally it updates status itself.
+func createTargetConfig(ctx context.Context, c TargetConfigController, recorder events.Recorder) (bool, error) {
+	var errors []error
+
+	_, _, err := ManageClientCABundle(ctx, c.configMapLister, c.kubeClient.CoreV1(), recorder, c.namespace)
+	if err != nil {
+		errors = append(errors, fmt.Errorf("%q: %v", "configmap/client-ca", err))
```
Comment (stevekuznetsov) [inline control-plane-pki-operator/targetconfigcontroller/targetconfigcontroller.go:72]: @stevekuznetsov update the name here

File: control-plane-pki-operator/targetconfigcontroller/targetconfigcontroller.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	if _, err := clienthelpers.UpdateHostedControlPlaneStatusCondition(ctx, condition, c.namespace, c.name, c.hypershiftClient); err != nil {
+		return true, err
+	}
+
+	return false, nil
+}
+
+func ManageClientCABundle(ctx context.Context, lister corev1listers.ConfigMapLister, client coreclientv1.ConfigMapsGetter, recorder events.Recorder, namespace string) (*corev1.ConfigMap, bool, error) {
+	requiredConfigMap, err := resourcesynccontroller.CombineCABundleConfigMaps(
+		resourcesynccontroller.ResourceLocation{Namespace: namespace, Name: manifests.TotalClientCABundle(namespace).Name}, // TODO: do we use a separate one, and converge by bringingbits over from the HCP controller?
```
Comment (stevekuznetsov) [inline control-plane-pki-operator/targetconfigcontroller/targetconfigcontroller.go:102]: @sjenning we should run through how we expect to do a migration here

Comment (stevekuznetsov) [reply control-plane-pki-operator/targetconfigcontroller/targetconfigcontroller.go:102]: (I left the HCP changes out of this changeset for now until we chat)

File: control-plane-pki-operator/operator.go (hunk 3 above)
Comment (deads2k) [reply control-plane-pki-operator/operator.go:24]: > @deads2k would like your guidance on parameterizing this - the controllercmd libarry does not allow for flags and will require hackery one way or ther other ...

we own the library.  Adjust as needed to do what you want.

File: control-plane-pki-operator/certrotationcontroller/certrotationcontroller.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+		UID:        hostedControlPlane.UID,
+	}
+
+	rotatorName := "CustomerAdminKubeconfigSigner"
+	certRotator := certrotation.NewCertRotationController(
+		rotatorName,
+		certrotation.RotatedSigningCASecret{
+			Namespace:     hostedControlPlane.Namespace,
+			Name:          manifests.CustomerSystemAdminSigner(hostedControlPlane.Namespace).Name,
+			Validity:      7 * rotationDay,
```
Comment (sjenning) [inline control-plane-pki-operator/certrotationcontroller/certrotationcontroller.go:80]: In my mind, I was thinking in the neighborhood of 1d validity and 12h rotation.

However, as long as we can adjust later, I'm not inclined to hold up the PR for consensus on this as long as we use something reasonable.

File: control-plane-pki-operator/operator.go (hunk 3 above)
Comment (stevekuznetsov) [reply control-plane-pki-operator/operator.go:25]: Just used `os.Getenv`, there's a little too many layers of complexity in `controllercmd` for me to grok for now.

**Comment by stevekuznetsov** [comment]
Added management of the new PKI controller to `hypershift-operator` - need to add an e2e now.

File: control-plane-pki-operator/certrotationcontroller/certrotationcontroller.go (hunk 1 above)
Comment (deads2k) [reply control-plane-pki-operator/certrotationcontroller/certrotationcontroller.go:56]: something similar is important to have here for testing with tighter timeframes.  I suggest keeping it.

File: control-plane-pki-operator/certrotationcontroller/certrotationcontroller.go (hunk 2 above)
Comment (deads2k) [reply control-plane-pki-operator/certrotationcontroller/certrotationcontroller.go:103]: username of "recognizeable prefix" + crypto-rand-number/uid.  Fully crypto-rand UID.

File: control-plane-pki-operator/certrotationcontroller/certrotationcontroller.go (hunk 1 above)
Comment (stevekuznetsov) [reply control-plane-pki-operator/certrotationcontroller/certrotationcontroller.go:56]: Moreso the question was - it looks like if you don't specify something custom in options, you get the dev duration. If you *do* specify something, though, you get a log warning about an unsupported value - that seems incorrect?

File: test/e2e/break_glass_credentials_test.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
+//go:build e2e
+// +build e2e
```
Comment (enxebre) [inline test/e2e/break_glass_credentials_test.go:2]: This validation is great but we don't need to spin up a new HC for it. We can let this be a validation of the createCluster test or the buildin logic that validate any new HC used in test.

File: control-plane-pki-operator/operator.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+func RunOperator(ctx context.Context, controllerContext *controllercmd.ControllerContext) error {
+	var namespace, name string
+	for env, target := range map[string]*string{
+		"HOSTED_CONTROL_PLANE_NAMESPACE": &namespace,
+		"HOSTED_CONTROL_PLANE_NAME":      &name,
+	} {
```
Comment (enxebre) [inline control-plane-pki-operator/operator.go:23]: fwiw there must only be a single HCP so onlny knowing the naespace is stricly required

File: control-plane-pki-operator/targetconfigcontroller/targetconfigcontroller.go [hunk 9]
```diff
@@ -90,23 +90,33 @@ func createTargetConfig(ctx context.Context, c TargetConfigController, recorder
 		Type:   "TargetConfigControllerDegraded",
 		Status: metav1.ConditionFalse,
 	}
-	if _, err := clienthelpers.UpdateHostedControlPlaneStatusCondition(ctx, condition, c.namespace, c.name, c.hypershiftClient); err != nil {
+	if _, err := clienthelpers.UpdateHostedControlPlaneStatusCondition(ctx, condition, c.hostedControlPlane.Namespace, c.hostedControlPlane.Name, c.hypershiftClient); err != nil {
```
Comment (enxebre) [inline control-plane-pki-operator/targetconfigcontroller/targetconfigcontroller.go:93]: In follow up we'll most likely want to bubble this up to the HC resource and include a check in validateHostedClusterConditions

File: test/e2e/break_glass_credentials_test.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+			if errors.IsNotFound(getErr) {
+				return false, nil
+			}
+			return getErr == nil, err
+		}); err != nil {
+			t.Errorf("client signer didn't become available: %v", err)
+		}
+
+		// Grab the break-glass client certificate
+		clientCertificate := manifests.CustomerSystemAdminClientCertSecret(hostedCluster.Namespace)
```
Comment (enxebre) [inline test/e2e/break_glass_credentials_test.go:54]: clientCertificate is fetched here but seems unused, shouldn't this be used in the kubeconfig below instead of the signer?

File: control-plane-pki-operator/operator.go (hunk 8 above)
Comment (stevekuznetsov) [reply control-plane-pki-operator/operator.go:23]: Yeah, I actually got hung up on this when I was learning the code-base ... if it's required to be a singleton, let's not write code that happens to look like it might work fine if there are N in the namespace - like the CPO ;)

File: control-plane-operator/controllers/hostedcontrolplane/manifests/pki.go [hunk 11]
```diff
@@ -170,6 +170,23 @@ func SystemAdminClientCertSecret(ns string) *corev1.Secret {
 	return secretFor(ns, "system-admin-client")
 }
 
+func CustomerSystemAdminSigner(ns string) *corev1.Secret {
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply control-plane-operator/controllers/hostedcontrolplane/manifests/pki.go:182]: I'll move it over

File: control-plane-pki-operator/targetconfigcontroller/targetconfigcontroller.go (hunk 9 above)
Comment (stevekuznetsov) [reply control-plane-pki-operator/targetconfigcontroller/targetconfigcontroller.go:93]: Would you prefer reporting in both spots? Or just HostedCluster instead of HostedControlPlane?

Comment (enxebre) [reply control-plane-pki-operator/targetconfigcontroller/targetconfigcontroller.go:93]: for cpo managed components the flow is usually to let them set their condition in the HCP then we bubble them up here https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go#L533.

I like it how you have it in the PR right now scoped to the HCP, we can later follow up to bubble it up.

Comment (stevekuznetsov) [reply control-plane-pki-operator/targetconfigcontroller/targetconfigcontroller.go:93]: Sounds good - please let me know what you expect for the roll-up (do we want all N conditions copied up, or an aggregate/summary?) and I'll add in the follow-up.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 12]
```diff
@@ -4818,3 +4859,169 @@ func isAPIServerRoute(hcluster *hyperv1.HostedCluster) bool {
...
+										FieldPath: "metadata.namespace",
+									},
+								},
+							},
+							{
+								Name:  "HOSTED_CONTROL_PLANE_NAME",
+								Value: hcp.Name,
+							},
+						},
+						Command: []string{"/usr/bin/control-plane-pki-operator"},
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:4927]: traditionally this would be a subcommand of the cpo, see https://github.com/openshift/hypershift/blob/3a42e77041535c8ac8012856d279bc782efcaf3c/control-plane-operator/main.go#L75
and 
https://github.com/openshift/hypershift/blob/3a42e77041535c8ac8012856d279bc782efcaf3c/hypershift-operator/controllers/hostedcluster/ignitionserver/ignitionserver.go#L554-L555

Comment (stevekuznetsov) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:4927]: @sjenning requested that it not be in this case

File: Makefile [hunk 13]
```diff
@@ -75,6 +75,10 @@ hypershift-operator:
 control-plane-operator:
 	$(GO_BUILD_RECIPE) -o $(OUT_DIR)/control-plane-operator ./control-plane-operator
 
+.PHONY: control-plane-pki-operator
```
Comment (enxebre) [inline Makefile:78]: With https://github.com/openshift/hypershift/pull/3193/commits/182804c9d0bffe49805e14b32df5563ac419aeee#r1395979845 then there's no need to adhoc build here since this will be a cpo subcommand

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 12 above)
Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:4927]: I assume is to keep library go controller code fully decoupled? sgmt whatever Seths preference.

Comment (stevekuznetsov) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:4927]: Yeah I think so

File: api/hypershift/v1beta1/hosted_controlplane.go [hunk 14]
```diff
@@ -267,6 +267,10 @@ type HostedControlPlaneStatus struct {
 	// Condition contains details for one aspect of the current state of the HostedControlPlane.
 	// Current condition types are: "Available"
 	// +optional
+	// +listType=map
```
Comment (sjenning) [inline api/hypershift/v1beta1/hosted_controlplane.go:270]: Why this change?  not sure if this will cause existing HCs to fail validation on update.

I would like to ensure this doesn't cause an issue on HO/CRD update.

File: Dockerfile [hunk 15]
```diff
@@ -11,6 +11,7 @@ COPY --from=builder /hypershift/bin/hypershift \
                     /hypershift/bin/hcp \
                     /hypershift/bin/hypershift-operator \
                     /hypershift/bin/control-plane-operator \
+                    /hypershift/bin/control-plane-pki-operator \
```
Comment (csrwng) [inline Dockerfile:14]: we also have `Containerfile.operator` and `fast.Dockerfile` that need to be updated.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 16]
```diff
@@ -2017,6 +2019,49 @@ func (r *HostedClusterReconciler) reconcileControlPlaneOperator(ctx context.Cont
...
+	controlPlanePKIOperatorRoleBinding := controlplaneoperator.PKIOperatorRoleBinding(controlPlaneNamespace.Name)
+	_, err = createOrUpdate(ctx, r.Client, controlPlanePKIOperatorRoleBinding, func() error {
+		return reconcileControlPlanePKIOperatorRoleBinding(controlPlanePKIOperatorRoleBinding, controlPlanePKIOperatorRole, controlPlanePKIOperatorServiceAccount)
+	})
+	if err != nil {
+		return fmt.Errorf("failed to reconcile controlplane PKI operator rolebinding: %w", err)
+	}
+
+	// Reconcile PKI operator cluster role
+	controlPlanePKIOperatorClusterRole := controlplaneoperator.PKIOperatorClusterRole(controlPlaneNamespace.Name)
```
Comment (csrwng) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2048]: In general, we don't bind any cluster roles to service accounts in the control plane namespace. If we need something from the infrastructure CR, the hypershift operator can get it and pass it either in the HCP or in args/env var to the PKI operator deployment.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 17]
```diff
@@ -4815,3 +4878,211 @@ func isAPIServerRoute(hcluster *hyperv1.HostedCluster) bool {
...
+			Verbs:     []string{"get"},
+		},
+		{ // for owner reference resolution
+			APIGroups: []string{"apps"},
+			Resources: []string{"replicasets"},
+			Verbs:     []string{"get"},
+		},
+		{ // for owner reference resolution
+			APIGroups: []string{"config.openshift.io"},
+			Resources: []string{"infrastructures"},
```
Comment (csrwng) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:5031]: I wouldn't think this works in a role

**Review by csrwng — COMMENTED** [review]
Awesome stuff. Some comments.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 17 above)
Comment (stevekuznetsov) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:5031]: Oops, forgot to remove here -it's in the cluster role below

File: api/hypershift/v1beta1/hosted_controlplane.go (hunk 14 above)
Comment (stevekuznetsov) [reply api/hypershift/v1beta1/hosted_controlplane.go:270]: This had to change because without it you can't PATCH or SSA one condition, you have to update all of them. I think you guys are doing UPDATE calls now so it won't impact you but at whatever point you go to PATCH or SSA to be more targeted, it will benefit you.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 16 above)
Comment (stevekuznetsov) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2048]: The `controllercmd` library uses a WATCH on this to determine how it should be doing leader election - in this case I don't think we can statically pass it in.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 18]
```diff
@@ -4815,3 +4887,206 @@ func isAPIServerRoute(hcluster *hyperv1.HostedCluster) bool {
...
+			Kind:      "ServiceAccount",
+			Name:      sa.Name,
+			Namespace: sa.Namespace,
+		},
+	}
+
+	return nil
+}
+
+func reconcileControlPlanePKIOperatorClusterRole(role *rbacv1.ClusterRole) error {
```
Comment (sjenning) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:5065]: I'm not understanding the need for this.  Why does the controller need to know the infrastructure of the MC?

Comment (stevekuznetsov) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:5065]: https://github.com/openshift/hypershift/pull/3193#discussion_r1399742409

File: control-plane-pki-operator/topology/detector.go [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+	hcp, err := client.HostedControlPlanes(namespace).Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("failed to get hosted control plane %s/%s: %q", namespace, name, err)
+	}
+
+	if hcp == nil {
+		return "", fmt.Errorf("got a nil hosted control plane for %s/%s", namespace, name)
+	}
+
+	switch hcp.Spec.InfrastructureAvailabilityPolicy {
```
Comment (enxebre) [inline control-plane-pki-operator/topology/detector.go:45]: the input that dictates hc controllers availability is hc.spec.ControllerAvailabilityPolicy.
InfrastructureAvailabilityPolicy express intent for "infra" running on guest cluster nodes e.g. ingress or registry controllers

File: control-plane-operator/controllers/hostedcontrolplane/pkioperator/reconcile.go [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+	mainContainer := hyperutil.FindContainer("control-plane-pki-operator", deployment.Spec.Template.Spec.Containers)
+	proxy.SetEnvVars(&mainContainer.Env)
+
+	deploymentConfig := config.DeploymentConfig{
+		Scheduling: config.Scheduling{
+			PriorityClass: config.DefaultPriorityClass,
+		},
+		SetDefaultSecurityContext: setDefaultSecurityContext,
+		AdditionalLabels: map[string]string{
+			config.NeedManagementKASAccessLabel: "true",
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/pkioperator/reconcile.go:106]: to pass e2e you'll need to add this component to the list of allowed ones in EnsureNetworkPolicies

Comment (stevekuznetsov) [reply control-plane-operator/controllers/hostedcontrolplane/pkioperator/reconcile.go:106]: I already did that and pass e2e? https://github.com/openshift/hypershift/pull/3193/commits/13094ab04b07808814f26cc6a3ca2c06d39ae43c#diff-cebde6b83b4b372162b2c6a22710a4ab0b1838576120535593a66a3a9bdfc170R767
